primitives/src/signatures/bls.rs

```diff
@@ -1,157 +1,316 @@
+//! // or a `SeedableRng` like `ChaChaRng` with seed generated from good randomness source.
+//! let (sk, pk) = BLSSignatureScheme::key_gen(&pp, &mut OsRng)?;
+//!
+//! let msg = "The quick brown fox jumps over the lazy dog";
+//! let sig = BLSSignatureScheme::sign(&pp, &sk, &msg, &mut OsRng)?;
+//! assert!(BLSSignatureScheme::verify(&pp, &pk, &msg, &sig).is_ok());
+//!
+//! # Ok::<(), Box<dyn std::error::Error>>(())
+//! ```
+//!
+//! ## Generating independent keys from the same IKM
+//!
+//! In case you want to keep the IKM for multiple key pairs, and potentially
+//! reconstruct them later on from IKM.
+//!
+//! ```
+//! use rand_core::{RngCore, OsRng};
+//! use sha2::{Sha256, Digest};
+//! use jf_primitives::signatures::{SignatureScheme, bls::BLSSignatureScheme};
+//!
+//! let pp = BLSSignatureScheme::param_gen::<OsRng>(None)?;
+//!
+//! // NOTE: in practice, please use [`zeroize`][zeroize] to wipe sensitive
+//! // key materials out of memory.
+//! let mut ikm = [0u8; 32]; // should be at least 32 bytes
+//! OsRng.fill_bytes(&mut ikm);
+//!
+//! let mut hasher = Sha256::new();
+//! hasher.update(b"MY-BLS-SIG-KEYGEN-SALT-DOM-SEP");
+//! let salt = hasher.finalize();
+//!
+//! let (sk1, pk1) = BLSSignatureScheme::key_gen_v5(&ikm, &salt, b"banking".as_ref())?;
+//! let (sk2, pk2) = BLSSignatureScheme::key_gen_v5(&ikm, &salt, b"legal".as_ref())?;
+//!
+//! let msg = "I authorize transfering 10 dollars to Alice";
+//! let sig = BLSSignatureScheme::sign(&pp, &sk1, &msg, &mut OsRng)?;
+//! assert!(BLSSignatureScheme::verify(&pp, &pk1, &msg, &sig).is_ok());
+//!
+//! let msg = "I agree to the Terms and Conditions.";
+//! let sig = BLSSignatureScheme::sign(&pp, &sk2, &msg, &mut OsRng)?;
+//! assert!(BLSSignatureScheme::verify(&pp, &pk2, &msg, &sig).is_ok());
+//!
+//! # Ok::<(), Box<dyn std::error::Error>>(())
+//! ```
+//!
+//! [zeroize]: https://github.com/RustCrypto/utils/tree/master/zeroize
 
 use super::SignatureScheme;
 use crate::{
     constants::{
-        BLS_SIG_KEY_SIZE, BLS_SIG_SIGNATURE_SIZE, BLS_SIG_VERKEY_SIZE, CS_ID_BLS_SIG_NAIVE,
+        BLS_SIG_COMPRESSED_PK_SIZE, BLS_SIG_COMPRESSED_SIGNATURE_SIZE, BLS_SIG_PK_SIZE,
+        BLS_SIG_SIGNATURE_SIZE, BLS_SIG_SK_SIZE, CS_ID_BLS_MIN_SIG,
     },
     errors::PrimitivesError,
 };
 
-use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, SerializationError};
+use ark_serialize::*;
 use ark_std::{
     convert::TryInto,
     format,
+    ops::{Deref, DerefMut},
     rand::{CryptoRng, RngCore},
+    vec::Vec,
 };
-
-use blst::BLST_ERROR;
+use blst::{min_sig::*, BLST_ERROR};
 use espresso_systems_common::jellyfish::tag;
 use tagged_base64::tagged;
+use zeroize::{Zeroize, Zeroizing};
 
-pub use blst::min_sig::{PublicKey, SecretKey, Signature};
-use zeroize::Zeroize;
-
-/// Newtype wrapper for a BLS Signing Key.
 #[tagged(tag::BLS_SIGNING_KEY)]
-#[derive(Clone, Debug, Zeroize)]
+#[derive(Clone, Debug, Default, Zeroize)]
```

Comment:
I'm not sure whether `Default` is a good idea on secret key. It might lead to users abusing it and causing insecure instantiations of the protocol.

Reply:
yeah, I hesitated also. Reason I added was that I saw the underlying type from `blst` have it.
I'll remove it.